### PR TITLE
Add vacuole upgrade to store double of only one compound

### LINKED ISSUE #4109: Add vacuole upgrade to double its storage but be only able to store one type
base work towards this has been done so it shouldn't be the hardest thing to make different compound types have different storage capacities, though this might have unexpected bugs and need fixing in many places as such I'm marking this as a difficult issue.

### PR DESCRIPTION
**Pull request by ender-titan1**
**Brief Description of What This PR Does**

This PR modifies CompoundBag to enable it to handle separate capacities for each compound,  as well as adding an upgrade to the vacuole to enable it to have double the capacity but only be able to handle one type of compound.

**Related Issues**

Closes #4109 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/microbe_stage/CompoundBag.cs
Comment (hhyyrylainen) [inline src/microbe_stage/CompoundBag.cs:32]: The internal details of how the capacity is stored should not be leaked. This should be made private. I created the `GetCapacityForCompound` method for a reason. And second problem is that if a mod for example adds more compounds or we add more compounds in a later update, this is so going to blow up and cause a game crash somewhere.

So here's my suggestion have a "nominal capacity" private field and a nullable dictionary of specific capacities also a private field. Then modify the GetCapacityForCompound to first check if a specific capacity exists for a compound, and if it does return it, otherwise return the nominal capacity. I think with those changes we'd have a design that I can think of being good enough.

Comment (ender-titan1) [reply src/microbe_stage/CompoundBag.cs:32]: Ok, on it

**Comment by ender-titan1** [comment]
I've implemented the requested changes, however, these changes created a problem where microbes would spawn without any initial compounds. I have fixed this in a somewhat hacky way and still do not know what exactly was causing the problem. I also wonder if `organellesCapacity` is still required in `Microbe.Internal.cs` since the capacity field had been made private.

**Comment by hhyyrylainen** [comment]
Regarding the bug with initial compounds, I've been converting that code over to a new system on my branch this week and I remember reading comments that the initial compound add used a direct add of compounds. That way it bypasses any normal limits or usefulness check. Could you modify your code to go back to that way of doing that to not need the workaround?

**Comment by ender-titan1** [comment]
So if I understand correctly, I should create a new method in `CompoundBag` to add compounds without checking capacity

**Comment by hhyyrylainen** [comment]
Can your work wait until my branch is merged as I already wrote this:
```
    public void AddInitialCompounds(IReadOnlyDictionary<Compound, float> compounds)
    {
        foreach (var entry in compounds)
        {
            if (!Compounds.TryGetValue(entry.Key, out var existingAmount))
            {
                Compounds[entry.Key] = entry.Value;
                continue;
            }

            float toAdd = entry.Value - existingAmount;

            if (toAdd > 0)
                Compounds[entry.Key] = existingAmount + toAdd;
        }
    }
```

doing it differently twice will lead to nice headache merge conflicts for me...

If I remember right didn't the initial compounds just use the `Compounds.Compounds` dictionary directly to set the values? That just goes past any limits etc.

**Comment by ender-titan1** [comment]
Yes, I can wait until that branch has been merged

**Comment by hhyyrylainen** [comment]
If wait until then you'll need to redo Microbe.Interior.cs, MicrobeAI.cs and ProcessSystem.cs changes (or solve very difficult merge conflicts as most of those files will be gone or moved).

I think something like:
```
// Temporary code before proper initial compounds set method is added to CompoundBag
Compounds.Compounds.TryGetValue(compound, out var existing);
Compounds.Compounds[compound] = existing + stuffToAdd;
```
Would be simple enough and this PR could move forward without having to wait a month, which is my maybe a realist estimate on when my work would be completed.

**Comment by ender-titan1** [comment]
I don't quite understand what you mean by "redo changes". Should I revert those files back to the original? And the code provided is meant to be inserted in the `SetInitialCompounds` method if I am to understand correctly.

**Comment by hhyyrylainen** [comment]
I'm just warning that as I'm making major changes, when those changes are merged to master that will lead to a large number of merge conflicts in your PR. So completing the functionality of the PR now with that temporary change will be less work for you. And then of course cleaning up the changes a bit to get this ready for merge, as I noticed at least a few places I'm going to comment on the design approach once this is functionally ready.

But if you are fine with that, we can get back to this PR in a month and then get this ready by fixing all the new merge conflicts with master.

File: src/microbe_stage/organelle_components/StorageComponent.cs
Comment (hhyyrylainen) [inline src/microbe_stage/organelle_components/StorageComponent.cs:42]: Is there a reason for using two variables? I'd just make one property `Compound? SpecializedFor` and make it so that null means no specialization and then a valid compound value is specialization for that compound.

File: src/late_multicellular_stage/MulticellularCreature.cs
Comment (hhyyrylainen) [inline src/late_multicellular_stage/MulticellularCreature.cs:493]: Looks like the indentation got messed up in this file.

File: src/microbe_stage/CompoundBag.cs
Comment (hhyyrylainen) [inline src/microbe_stage/CompoundBag.cs:16]: I don't see why this would need to be a field. Can't this be a property? That avoid having a "useless" manually written setter for the nominal capacity.

File: src/microbe_stage/CompoundBag.cs
Comment (hhyyrylainen) [inline src/microbe_stage/CompoundBag.cs:49]: I think this comment needs to elaborate that this is the capacity for all compounds that don't have a specific capacity set.

File: src/microbe_stage/Microbe.Interior.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Interior.cs:4]: I'm pretty sure this is a mistake, the microbe code should not be doing any HTML parsing...

**Comment by hhyyrylainen** [comment]
I think you hit a bunch of warnings that were hiding underneath the surface on master some time ago. I think those should go away if you merge master into your branch.

File: src/microbe_stage/organelle_components/StorageComponent.cs
Comment (hhyyrylainen) [inline src/microbe_stage/organelle_components/StorageComponent.cs:41]: This doesn't sound grammatically correct to me, is it? To me `SpecializedFor` would sound grammatically correct when reading code aloud that uses this property. Would that be better?

File: src/microbe_stage/CompoundBag.cs
Comment (hhyyrylainen) [inline src/microbe_stage/CompoundBag.cs:24]:
```suggestion
    ///   Specifies the default capacity for all compounds that do
```
As this variable doesn't "store" the data just specifies how much something can be stored elsewhere.

File: src/microbe_stage/CompoundBag.cs
Comment (hhyyrylainen) [inline src/microbe_stage/CompoundBag.cs:42]:
```suggestion
    ///   Returns the capacity this bag has for storing the compound if it is useful, otherwise 0
```
I'd omit the internal details, especially the non `<see ...` type reference to a private variable.

File: src/microbe_stage/Microbe.Interior.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Interior.cs:1350]: I'm noticing that you are wrapping the documentation text really tightly? Is there a reason? The documentation comments should be filled all the way to column 120.

File: src/microbe_stage/Microbe.Interior.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Interior.cs:1386]: This doesn't remove special capacities that are now gone from the cell (for example if the player switches a vacuole from glucose to ammonia, they get the extra capacity for both as the old value is never seemingly removed). There needs to be a method in the compound bag to reset the special capacities before they are re-applied to ensure no outdated values exist.

File: src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:40]: I don't see you modifying the shown choices so you should be able to just use the returned IReadOnlyList directly here instead of taking a copy of the data.

File: src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:52]: Don't the other upgrade GUIs have the default value stored in Constants? This should also be in constants and named something like `DEFAULT_VACUOLE_UPGRADE_FOR_COMPOUND = "glucose"`.

File: src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:65]: Shouldn't this be `defaultCompoundIndex`? to not have to find the same index again.

**Comment by hhyyrylainen** [comment]
A change was just merged that requires adding a comparison method for the upgrade data: https://github.com/Revolutionary-Games/Thrive/pull/4463 so I suspect that that'll need a small tweak to this PR to make sure this compiles (with latest master).

File: src/microbe_stage/Microbe.Interior.cs
Comment (ender-titan1) [reply src/microbe_stage/Microbe.Interior.cs:1380]: Regarding this, the feature still works as intended. This is because `additionalCompoundCapacities` for in your example glucose is set to zero when the capacity is begin calculated. This means that while the entry exists in the dictionary, the value is zero. I have also play-tested this and the feature works as intended.

File: src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs
Comment (ender-titan1) [reply src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:40]: Forgive me but I do not understand, `GetAllCompounds` returns a dictionary and the `Values` property returns a ValueCollection that I cannot convert to a list. I am a bit confused and ask for clarification.

File: src/microbe_stage/Microbe.Interior.cs
Comment (hhyyrylainen) [reply src/microbe_stage/Microbe.Interior.cs:1380]: Ah, you mean with `UpdateCapacity`? I guess that is true, though with the way I'm refactoring the microbe stage the "organelle removed" callback will be entirely removed, so could you futureproof your code by doing that my suggestion? That way I don't accidentally introduce this bug when trying to fix merge conflicts with my branch.

File: src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs
Comment (hhyyrylainen) [reply src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:40]: My mistake, I thought GetAllCompounds would return a `IReadOnlyList<Compound>`. Alright then this is a fine approach without any changes.

File: RevolutionaryGamesCommon
Comment (hhyyrylainen) [inline RevolutionaryGamesCommon:1]: Does this PR accidentally downgrade the common module version? It seems like this is moving it back to version `f1b5b669dbce77e203bc001bc662d2d549b67a40` even though it was newer version before on the master branch.

File: src/microbe_stage/Microbe.Interior.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Interior.cs:1353]:
```suggestion
    /// <param name="negative">Should be true if the organelle is being removed, otherwise false</param>
```

**Comment by hhyyrylainen** [comment]
Just those two minor things I still noticed (well one is possibly not minor as downgrading the common submodule is a bit of real problem). Once those are fixed I think I'll just merge this, unless I can manage to get someone to playtest these changes.

**Comment by buckly90** [comment]
Seems to work perfectly well, with one unfortunate exception.

Iron storage is not being increased as it should be, and the upgrade has no appearant effect.

**Comment by ender-titan1** [comment]
Acknowledged, I will see what is causing the iron capacity not to increase, and I will try to resolve the submodule problem as soon as i can.

**Comment by ender-titan1** [comment]
I've fixed the iron issue, however, I do not know how to fix the submodule problem. I've never worked with submodules in the past and trying to update them using git bash creates an error in the code, so I really don't know what I am supposed to do

**Comment by hhyyrylainen** [comment]
> I've never worked with submodules in the past and trying to update them using git bash creates an error in the code, so I really don't know what I am supposed to do

Well that explains it. You've pulled in new code and then committed *all changes* which incorrectly picked the old submodule version. Git pull or branch switching do not update submodules, you need to run `git submodule update` to update submodules. You can probably fix things by:
```
cd RevolutionaryGamesCommon
git checkout 3a737999e08a88cd728067a7f4be695ef1d65f10
cd ..
git add ...
git commit ...
```

to create a new commit that sets the submodule version to what it is on master.

**Review by hhyyrylainen — APPROVED** [review]
Code changes look good now. Also the fix to the iron probably fixed all vacuole configuration issues from the underlying problem of the Godot option button not having the same items as the internal compounds list.

**Comment by hhyyrylainen** [comment]
@buckly90 do you want to re-test this to make sure the iron bug is fixed? That's the last potential thing to test before merging this. At the latest I'll merge this in two days to make sure this makes it into the BOTD.

**Comment by buckly90** [comment]
Looks great, splendid job!

**Pull request merged**
